show-url.1
.TH SHOW\-URL 1 "2006 June 17"
.ds su \fIshow\-url\fP
.ds Su \fIShow\-url\fP
.UC 4
.SH NAME
show\-url \- show a URL in netscape
.SH SYNOPSIS
.nf
show\-url [OPTION]... [filename]
show\-url [OPTION]... \-u url
.fi
.SH OPTIONS
\*(Su may be invoked with the following command-line option:
.TP
.B \-b \fIbrowser\fP
.TP
.BI \-\-browser= browser
Override the value of the BROWSER environment variable.  See
DESCRIPTION below for details.
.TP
.B \-d
.TP
.B \-\-debug
.TP
.BI \-\-debug= level
Set the debug level.  Without a level, this will increase the
debugging level by one.
.TP
.B \-u \fIurl\fP
.TP
.BI \-\-url= url
Display \fIurl\fP rather than reading one from a file.  If \fIurl\fP
names a readable file then a corresponding file url will be
constructed.  Otherwise \fIurl\fP will be passed to the browser
without modification.
.TP
.B \-v
.TP
.B \-\-version
Print out the show\-url version and exit.
.B \-h
.TP
.B \-\-help
Print a usage message briefly summarizing these command-line options
and then exit.
.SH DESCRIPTION
\*(Su is deigned to be a handler for the text/uri\-list MIME type.  It
attempts to display the URL listed in the first line of \fIfilename\fP
in a web browser.  If \fIfilename\fP is "\-" or neither \fIfilename\fP
or \fIurl\fP is given then the URL is read from stdin.
.PP
The command used to invoke the web browser is specified by the BROWSER
environment variable, which is a colon-separated list of Bourne shell
expressions.  Each expression may optionally contain `%s' which will
be replaced with the URL when executing the expression.  `%\fIc\fP'
where \fIc\fP is not `s' will be replaced simply by \fIc\fP, thus the
expression may use `%%s' to generate a literal `%s' and `%:' for a
literal `:'.  If the expression does not contain a `%s' then the URL
is appended to the end of the expression.
.PP
Shell expressions are executed until one exits with a zero status
code.
.PP
Characters in the URL which
are meaningful to the shell are escaped to prevent malicious URLs from
doing naughty things.
.SH EXAMPLES
Here are some example values for the BROWSER variable:
.TP 10
.B
w3m
This will open the URL in w3m in the current terminal window.
.TP
.B
mozilla \-raise \-remote "openURL(%s,new\-window)"
Requests that the running mozilla process open the URL in a new
window.
.TP
.B
mozilla \-raise \-remote "openURL(%s,new\-window)":w3m
A combination of the above two.  If mozilla cannot open the window
because it isn't running or there's not even an X session, then w3m is
invoked to display the URL.
.TP
.B
netscape \-raise \-remote "openURL(%s,new\-window)":lynx
This is effectively the same as the previous example except that
netscape replaces mozilla as the graphical browser, and lynx replaces
w3m as the text-based one.
.SH ENVIRONMENT
.TP 10
.SB BROWSER
See DESCRIPTION above for an explanation of this variable.  If BROWSER
is not set no browser is specified on the command-line then the
expression given in the last example above is used.
.SH SEE ALSO
.BR xtickertape (1),
.na
http://tuxedo.org/~esr/BROWSER/index.html
.SH BUGS
E-mail bug reports to
.BR phelps@pobox.com .
Be sure to include the word "show\-url" somewhere in the "Subject:"
field.
.SH AUTHORS
\*(Su was written by Ted Phelps <phelps@pobox.com>.  Additional help
was provided by David Arnold <davida@pobox.com>.  Eric S. Raymond
<esr@thyrsus.com> is responsible for the BROWSER environment variable
format.


NEWS
Overview of changes in xtickertape-2.0

* Plugged 3 memory leaks found by valgrind, and removed a zero-length
  malloc which would result in an abort on platforms which return NULL.

* Added a workaround for a Motif bug so that the group menu reflects
  the current selection even before the user clicks on it for the first time.

* Changed the default attachment type from x-elvin/url to the standard
  text/uri-list for better interoperation with other tickertape clients.

* The show-url program can now read a URL from stdin.

Overview of changes in xtickertape-2.0rc1

* Notification format upgraded to conform to version 3.1 of the
  tickertape chat protocol.

* It's now finally possible to reload a groups file in which the keys
  for a group have changed.  It's also possible to reload the keys
  file or all config files (with a SIGHUP).

* It is now possible to reply to a message in a 'no menu' group.

* Fixed a thinko in the metamail invocation which caused the first
  argument to be lost and thus metamail would cause it to hang.

Overview of changes in xtickertape-2.0b2

* Fixed a bug that prevented the connection status from being updated
  properly.

* Added the `-u' option to show-url command, making it possible to
  specify the URL on the command-line rather than writing it to a file
  first.

* Changed the metamail invocation to use execlp instead of execl,
  which means that the PATH environment variable is consulted.  The
  corresponding app-default has changed name from [Mm]etamailPath to
  simply [Mm]etamail.

* Binary package support for Redhat, Mandrake and Debian.

* Fixed a counting error that could cause xtickertape to crash when
  trying to display certain kinds of MIME attachments.  Also rewriting
  the MIME headers before sending them to metamail, which doesn't cope
  well with any end-of-line convention but LF.

Overview of changes in xtickertape-2.0b1

* Messages containing non-ASCII characters are now sent and displayed
  correctly.

* Reworked the way dragging is handled by the Scroller widget so that
  you can perform actions on a message while dragging the scroller
  without cancelling the drag.

* Added a `send history' which records the last few messages sent by
  the current user so that they can go back and send them to other
  groups or fix the spelling.

* New command-line options: -T ticker-dir and -k keys-dir.  The former
  overrides the TICKERDIR environment variable, and the latter
  specifies the root path of a key file with a relative pathname.  If
  keys-dir is not set then it defaults to the directory containing the
  keys file.

Overview of changes in xtickertape-2.0a8

* The usenet file now supports backslashes as escape characters.  You
  can now subscribe to URLs.

* It is now possible to reply to a message in a different group by
  first selecting the message to which you are replying and then
  selecting the group to send it to.  To clear the current selection,
  click on it in the history.

* The show-url command has been completely rewritten to use the
  BROWSER environment variable as defined by ESR.

* xtickertape can now build against verion 4.1b4 of libelvin.

Overview of changes in xtickertape-2.0a7

* The path to metamail is now an X resource.

* The application now refuses to run if the app-defaults isn't found
  or has a different version number than the executable, and prints an
  error message which should hopefully help the user to address the
  problem.  [Thanks Mark White]

* Added a way to set the idle-period to test the connection after some
  period of inactivity.

Overview of changes in xtickertape-2.0a6

* The Send button is now disabled until there is at least one
  character in the `Text' field.

* Group notifications now conform to both the old message format and
  the v3.0 draft specification.

* Created a new widget (class History) to display the last few
  messages.  The length of the history can be specified by setting
  the XTickertape.history.messageCapacity X resource.  See the man
  page for additional History widget resources.

* Replaced tool-tips with a status bar.

* Changed the application class from `Tickertape' to `XTickertape'.

* No longer redirecting the output of `metamail' to /dev/null, and
  using the `-x' option so that it doesn't try to talk to the tty.
  [Thanks David Leonard]

* Also passing the mime args to metamail through a pipe instead of a
  temporary file.  [Thanks David Leonard]

* Fixed a bug in the way the icon was created.  Now the big `e' will
  be red even on 8-bit displays.

* The scroller can now operate without an offscreen pixmap.  This
  enables hardware optimizations with some graphics cards, but results 
  in more traffic between the client and server.  The offscreen pixmap 
  option can still be enabled by setting the X resource
  XTickertape.scroller.usePixmap to True.

* Message replacement now works!

* Updated to use the elvin-4 client library instead of the elvin-3
  client library.

* Using elvin's sha1digest function instead of crypt, and encrypting
  the entire message-id instead of just the host id to make it
  reasonably difficult to subvert the sender's anonymity.
  [Thanks Ian Lister]

* Using the elvin-4 `opaque' type to represent an attachment rather
  than a `string'.  This makes it possible to display images, play
  sounds or view anything else you have a decoder for via tickertape.
  Of course, getting those things attached is somewhat tricky.
  [Thanks Martin Wanicki]

* The `show-menu' action now maps the control panel window so that it
  will be visible even if it has been iconified [Thanks David Leonard]

Overview of changes in xtickertape-1.6.6

* Minimum size of the control panel now takes into account the height
  of the horizontal scrollbar on the history list.

* Minor configure fix for machines which don't have a separate include
  file for crypt() (thanks David Leonard for pointing out that BSD
  machines don't have <crypt.h>).


show-url.c
/***********************************************************************

  Copyright (C) 2002-2004 by Mantara Software (ABN 17 105 665 594).
  All Rights Reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:

   * Redistributions of source code must retain the above
     copyright notice, this list of conditions and the following
     disclaimer.

   * Redistributions in binary form must reproduce the above
     copyright notice, this list of conditions and the following
     disclaimer in the documentation and/or other materials
     provided with the distribution.

   * Neither the name of the Mantara Software nor the names
     of its contributors may be used to endorse or promote
     products derived from this software without specific prior
     written permission. 

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
   FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
   REGENTS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
   INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
   BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
   CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
   POSSIBILITY OF SUCH DAMAGE.

***********************************************************************/

#ifndef lint
static const char cvsid[] = "$Id: show-url.c,v 1.17 2006/06/17 10:23:11 phelps Exp $";
#endif /* lint */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <stdio.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_STDARG_H
#include <stdarg.h>
#endif

#ifndef WEXITSTATUS
#define WEXITSTATUS(stat_val) ((unsigned int)(stat_val) >> 8)
#endif

/* Environment variables */
#define ENV_BROWSER "BROWSER"
#define DEF_BROWSER "netscape -raise -remote \"openURL(%s,new-window)\":lynx"
#define MAX_URL_SIZE 4095
#define INIT_CMD_SIZE 128
#define FILE_URL_PREFIX "file://"

/* Options */
#define OPTIONS "b:dhu:v"

#if defined(HAVE_GETOPT_LONG)
/* The list of long options */
static struct option long_options[] =
{
    { "browser", required_argument, NULL, 'b' },
    { "url", required_argument, NULL, 'u' },
    { "debug", optional_argument, NULL, 'd' },
    { "version", no_argument, NULL, 'v' },
    { "help", no_argument, NULL, 'h' },
    { NULL, no_argument, NULL, '\0' }
};
#endif

/* A table converting a number into a hex nibble */
static char hex[] =
{
    '0', '1', '2', '3', '4', '5', '6', '7', '8',
    '9', 'a', 'b', 'c', 'd', 'e', 'f'
};

/* Characters which should be escaped */
static char no_esc[] =
{
    2, 1, 2, 0,  1, 0, 1, 1,  1, 2, 1, 0,  2, 0, 0, 0,  /* 0x20 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 1,  1, 0, 1, 1,  /* 0x30 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x40 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  1, 0, 1, 0,  /* 0x50 */
    2, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x60 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  1, 0, 0, 2,  /* 0x70 */
};

/* Characters which need to be escaped inside double-quotes */
static char dq_esc[] =
{
    2, 2, 2, 0,  1, 0, 0, 0,  0, 2, 0, 0,  2, 0, 0, 0,  /* 0x20 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x30 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x40 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x50 */
    2, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x60 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 2,  /* 0x70 */
};

/* Characters which need to be escaped inside single-quotes */
static char sq_esc[] =
{
    2, 0, 2, 0,  0, 0, 0, 2,  0, 2, 0, 0,  2, 0, 0, 0,  /* 0x20 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x30 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x40 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x50 */
    2, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  /* 0x60 */
    0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 2,  /* 0x70 */
};

/* The program name */
static char *progname = NULL;

/* The buffer used to construct the command */
static char *cmd_buffer = NULL;

/* The next character in the command buffer */
static size_t cmd_index = 0;

/* The end of the command buffer */
static size_t cmd_length = 0;

/* The verbosity level */
static int verbosity = 0;


/* Debugging printf */
void dprintf(int level, const char *format, ...)
{
    va_list ap;

    /* Bail if the statement is below the debugging threshold */
    if (verbosity < level) {
        return;
    }

    /* Do the varargs thing */
    va_start(ap, format);
    vfprintf(stderr, format, ap);
    va_end(ap);

    fflush(stderr);
}

/* Print out a usage message */
static void usage()
{
    fprintf(stderr,
	    "usage: %s [OPTION]... filename\n"
	    "usage: %s [OPTION]... -u URL\n"
	    "  -b browser,     --browser=browser\n"
	    "  -d,             --debug[=level]\n"
	    "  -u,             --url=url\n"
	    "  -v,             --version\n"
	    "  -h,             --help\n", progname, progname);
}

/* Appends a character to the command buffer */
static void append_char(int ch)
{
    /* Make sure there's enough room */
    if (! (cmd_index < cmd_length))
    {
	dprintf(3, "growing buffer\n");

	/* Double the buffer size */
	cmd_length *= 2;
	if ((cmd_buffer = (char *)realloc(cmd_buffer, cmd_length)) == NULL)
	{
	    perror("realloc() failed");
	    exit(1);
	}
    }

    cmd_buffer[cmd_index++] = ch;
}

/* Appends an URL to the command buffer, apply escapes as appropriate */
static void append_url(char *url, int quote_count)
{
    char *point;
    char *esc_table;

    /* Figure out which escape table to use */
    switch (quote_count)
    {
	case 0:
	{
	    esc_table = no_esc;
	    break;
	}

	case 1:
	{
	    esc_table = sq_esc;
	    break;
	}

	case 2:
	{
	    esc_table = dq_esc;
	    break;
	}

	default:
	{
	    /* Funny kind of quoting */
	    abort();
	}
    }

    point = url;
    while (1)
    {
	int ch = *point;
	int esc_type;

	/* Watch for the end of line/input */
	if (ch == '\0' || ch == '\n')
	{
	    return;
	}

	/* Figure out how to escape this character */
	esc_type = (ch < 32 || ch > 127) ? 2 : esc_table[ch - 32];
	switch (esc_type)
	{
	    /* No escape required */
	    case 0:
	    {
		append_char(ch);
		break;
	    }

	    /* Escape with a backslash */
	    case 1:
	    {
		append_char('\\');
		append_char(ch);
		break;
	    }

	    /* URL escape */
	    case 2:
	    {
		append_char('%');
		append_char(hex[(ch >> 4) & 0xf]);
		append_char(hex[ch & 0xf]);
		break;
	    }

	    /* Trouble */
	    default:
	    {
		abort();
	    }
	}

	point++;
    }
}

/* Invoke the browser on the given URL */
char *invoke(char *browser, char *url)
{
    int did_subst = 0;
    char *point = browser;
    int quote_count = 0;

    /* Reset the buffer */
    cmd_index = 0;

    /* Copy from the browser string */
    while (1)
    {
	int ch = *point;

	switch (ch)
	{
	    /* Watch for the end of the browser string */
	    case '\0':
	    case ':':
	    {
		int status;

		/* Insert the URL if we haven't done so already */
		if (! did_subst)
		{
		    append_char(' ');
		    append_url(url, quote_count);
		}

		/* Null-terminate the command */
		append_char('\0');

		/* Invoke the command */
		dprintf(1, "exec: %s\n", cmd_buffer);

		if ((status = system(cmd_buffer)) < 0)
		{
		    perror("fork() failed");
		    exit(1);
		}

		/* If successful return NULL */
		if (WEXITSTATUS(status) == 0)
		{
		    dprintf(2, "ok\n");
		    return NULL;
		}

		dprintf(2, "failed: %d\n", WEXITSTATUS(status));
		return ch == '\0' ? point : point + 1;
	    }

	    /* Watch for double-quotes */
	    case '"':
	    {
		/* Toggle double-quotes if appropriate */
		if (quote_count == 2)
		{
		    quote_count = 0;
		}
		else if (quote_count == 0)
		{
		    quote_count = 2;
		}

		append_char(ch);
		break;
	    }

	    /* Watch for single-quotes */
	    case '\'':
	    {
		/* Toggle single-quotes if appropriate */
		if (quote_count == 1)
		{
		    quote_count = 0;
		}
		else if (quote_count == 0)
		{
		    quote_count = 1;
		}

		append_char(ch);
		break;
	    }

	    /* Watch for %-escapes */
	    case '%':
	    {
		ch = point[1];

		/* Watch for the URL substitution */
		if (ch == 's')
		{
		    append_url(url, quote_count);

		    /* Skip ahead */
		    did_subst = 1;
		    point++;
		    break;
		}

		/* Watch for odd EOF */
		if (ch == '\0')
		{
		    append_char('%');
		    break;
		}

		/* Otherwise drop the initial % */
		append_char(ch);
		point++;
		break;
	    }

	    default:
	    {
		append_char(ch);
		break;
	    }
	}

	point++;
    }
}


/* Read the URL from a file and use it to invoke a browser */
int main(int argc, char *argv[])
{
    char *browser;
    char buffer[MAX_URL_SIZE + 1];
    struct stat statbuf;
    char *url = NULL;
    char *filename = NULL;
    FILE *file;
    size_t length;
    char *point;

    /* Extract the program name from argv[0] */
    if ((progname = strrchr(argv[0], '/')) == NULL)
    {
	progname = argv[0];
    }
    else
    {
	progname++;
    }

    /* Get the browser string from the environment */
    if ((browser = getenv(ENV_BROWSER)) == NULL)
    {
	browser = DEF_BROWSER;
    }

    while (1)
    {
	int choice;

#if defined(HAVE_GETOPT_LONG)
	choice = getopt_long(argc, argv, OPTIONS, long_options, NULL);
#else
	choice = getopt(argc, argv, OPTIONS);
#endif

	/* End of options? */
	if (choice < 0)
	{
	    break;
	}

	/* Which option is it? */
	switch (choice)
	{
	    /* --browser= or -b */
	    case 'b':
	    {
		browser = optarg;
		break;
	    }

	    /* --debug= or -d */
	    case 'd':
	    {
		if (optarg == NULL)
		{
		    verbosity++;
		}
		else
		{
		    verbosity = atoi(optarg);
		}

		break;
	    }

	    /* --help or -h */
	    case 'h':
	    {
		usage();
		exit(0);
	    }

	    /* --url= or -u */
	    case 'u':
	    {
		/* Determine if we should view a local file or a remote one */
		if (stat(optarg, &statbuf) < 0)
		{
		    dprintf(2, "%s: unable to stat file: %s\n", progname, optarg);
		    url = strdup(optarg);
		}
		else if ((statbuf.st_mode & S_IRUSR) == 0)
		{
		    dprintf(2, "%s: unable to read file: %s\n", progname, optarg);
		    url = strdup(optarg);
		}
		else
		{
		    dprintf(3, "%s: creating file URL: %s\n", progname, optarg);
		    if (optarg[0] == '/')
		    {
			length = sizeof(FILE_URL_PREFIX) + strlen(optarg);
			if ((url = (char *)malloc(length)) == NULL)
			{
			    perror("malloc() failed");
			    exit(1);
			}

			snprintf(url, length, FILE_URL_PREFIX "%s", optarg);
		    }
		    else
		    {
			/* Construct an absolute path */
			if (getcwd(buffer, MAX_URL_SIZE) == NULL)
			{
			    perror("Unable to determine current directory");
			    exit(1);
			}

			length = sizeof(FILE_URL_PREFIX) + strlen(buffer) + 1 + strlen(optarg);
			if ((url = (char *)malloc(length)) == NULL)
			{
			    perror("malloc() failed");
			    exit(1);
			}

			snprintf(url, length, FILE_URL_PREFIX "%s/%s", buffer, optarg);
		    }
		}

		dprintf(2, "%s: raw URL: %s\n", progname, url);
		break;
	    }

	    /* --version or -v */
	    case 'v':
	    {
		printf(PACKAGE " show-url version " VERSION "\n");
		exit(0);
	    }

	    /* Unsupported option */
	    case '?':
	    {
		usage();
		exit(1);
	    }

	    /* Trouble */
	    default:
	    {
		abort();
	    }
	}
    }

    /* If no URL or filename provided then read from stdin */
    if (url == NULL)
    {
	if (optind < argc)
        {
            /* Get the filename */
            filename = argv[optind++];

            /* Treat '-' as a synonym for stdin */
            if (strcmp(filename, "-") == 0) {
                filename = NULL;
            }
        }
    }

    /* Make sure there are no more arguments */
    if (optind < argc)
    {
	usage();
	exit(1);
    }

    /* Extract the URL from the file */
    if (url == NULL) {
        if (filename == NULL)
        {
            dprintf(3, "%s: reading URL from stdin\n", progname);
            filename = "<stdin>";
            file = stdin;
        }
        else 
        {
            dprintf(3, "%s: reading URL from %s\n", progname, filename);
            
            /* Read the URL from the file */
            if ((file = fopen(filename, "r")) == NULL)
            {
                perror("Unable to open URL file");
                exit(1);
            }
        }

        /* Read up to MAX_URL_SIZE bytes of it */
        length = fread(buffer, 1, MAX_URL_SIZE, file);
        buffer[length] = 0;

        dprintf(2, "%s: raw URL: %s\n", progname, buffer);

        /* Clean up */
        fclose(file);

        /* Duplicate its contents */
        if ((url = strdup(buffer)) == NULL)
        {
            perror("malloc() failed");
            exit(1);
        }

        strcpy(url, buffer);
    }

    /* Initialize the command buffer */
    cmd_length = INIT_CMD_SIZE;
    if ((cmd_buffer = (char *)malloc(cmd_length)) == NULL)
    {
	perror("malloc() failed");
	exit(1);
    }

    /* Attempt to open a browser */
    point = browser;
    while (*point != '\0')
    {
	point = invoke(point, url);
	if (point == NULL)
	{
	    exit(0);
	}
    }

    /* Clean up */
    free(url);
    exit(1);
}
